volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+                name="pdbutil", component=pdbutil.PDBUtility, version=(1, 0, 0)
+            ),
+            requirements.ChoiceRequirement(
+                name="source",
+                description="Where to resolve symbols.",
+                choices=["kernel", "processes"],
+                optional=False,
+            ),
+            requirements.StringRequirement(
+                name="module",
+                description='Module in which to resolve symbols. Use "ntoskrnl.exe" to resolve in the base kernel executable.',
+                optional=False,
+            ),
+            requirements.StringRequirement(
+                name="symbol",
+                description="Symbol name to resolve",
+                optional=True,
+            ),
+            requirements.IntRequirement(
+                name="address",
+                description="Address of symbol to resolve",
+                optional=True,
+            ),
+        ]
+
+    @staticmethod
+    def _get_pefile_obj(
+        context: interfaces.context.ContextInterface,
+        pe_table_name: str,
+        layer_name: str,
+        base_address: int,
+    ) -> Optional[pefile.PE]:
+        """
+        Attempts to pefile object from the bytes of the PE file
+
+        Args:
+            pe_table_name: name of the pe types table
+            layer_name: name of the process layer
+            base_address: base address of the module
+
+        Returns:
+            the constructed pefile object
+        """
+        pe_data = io.BytesIO()
+
+        try:
+            dos_header = context.object(
+                pe_table_name + constants.BANG + "_IMAGE_DOS_HEADER",
+                offset=base_address,
+                layer_name=layer_name,
+            )
+
+            for offset, data in dos_header.reconstruct():
+                pe_data.seek(offset)
+                pe_data.write(data)
+
+            pe_ret = pefile.PE(data=pe_data.getvalue(), fast_load=True)
+
+        except exceptions.InvalidAddressException:
+            pe_ret = None
+
+        return pe_ret
+
+    @staticmethod
+    def range_info_for_address(
+        ranges: ranges_type, address: int
+    ) -> Optional[range_type]:
+        """
+        Helper for getting the range information for an address
+        """
+        for start, size, filepath in ranges:
+            if start <= address < start + size:
+                return start, size, filepath
+
+        return None
+
+    @staticmethod
+    def filepath_for_address(ranges: ranges_type, address: int) -> Optional[str]:
+        """
+        Helper to get the file path for an address
+        """
+        info = PESymbols.range_info_for_address(ranges, address)
+        if info:
+            return info[2]
+
+        return None
+
+    @staticmethod
+    def filename_for_path(filepath: str) -> str:
+        """
+        Consistent way to get the filename
+        """
+        return filepath.split("\\")[-1]
```

Comment:
The VADS will have paths like:

```
\Windows\system32\kernel32.dll
```

The function gets the kernel32.dll part

Reply:
Ok, sorry, I read it as `strip` not `split` for some reason.  I don't suppose [`os.path.basename`](https://docs.python.org/3/library/os.path.html#os.path.basename) would do this?  Just seems better to let python maintain that little bit of code in case it turns out to be complex in some situations...

Reply:
No, because `basename` is really annoying and operates based on the OS Python is running on, so if you run on Linux or Mac then it only understands / as directories. From my Mac:

```
>>> import os
>>> os.path.basename("\\Windows\\System32\\kernel32.dll")
'\\Windows\\System32\\kernel32.dll'
```

It would only work if people were running Vol3 on Windows... its very weird behaviour

Reply:
Ugh, that's disappointing... oh well.  I guess just do up the docstring to explain what it's doing (basename rather than filename, or that's it's cutting off the path or something to make it slightly easier to grok).

Reply:
After seeing this comment, Gus showed me the ntpath module so now this function is much cleaner. ntpath dynamically determines \ vs / depending on the host.